components/server/src/gitlab/gitlab-context-parser.ts

```diff
@@ -214,29 +214,31 @@ export class GitlabContextParser extends AbstractContextParser implements IConte
         const possibleBranches = await this.gitlabApi.run<GitLab.Branch[]>(user, async g => {
```

Comment:
why so complicated? 
there is https://docs.gitlab.com/ee/api/branches.html#get-single-repository-branch

Reply:
The problem is, that we don't know what of the segments is part of the branch and what is part of the file tree.

For example this context URL:

https://gitlab.com/gp-test-group/gp-test-subgroup/gp-test-project-in-subgroup/-/tree/corneliusludmann/important-issue-1/test-dir-1

- `corneliusludmann/important-issue-1` is the branch and
-  `test-dir-1` is the file path.

But we cannot know this having just the context URL. It could also be that:
- `corneliusludmann` is the branch and
- `important-issue-1/test-dir-1` is the file path.

We don't know this until we check for all possible branches that start with the first segement.

Reply:
I believe this is the root cause of the issue here. A lucky shot revealed that some/lot of projects are structuring their branches in `feature/9999_something` patterns. 
In L.212 we're searching for `^feature` branches, which in many cases will be a large set of branches. Unfortunately, there is pagination in the GL REST API unconsidered here at: https://docs.gitlab.com/ee/api/#pagination
So, we always see only some of the `feature*` branches (maybe just 20.)

Reply:
That is an interesting thought! Indeed, pagination could be the problem here. :bulb:

Reply:
@corneliusludmann thanks for the explanation. I forgot about this dilemma.
Still, in that case we need a retry strategy with `get-single-repository-branch` calls and path prefix' in in ascending order, i.e. call it for `seg1`, `seg1/seg2`, `seg1/seg2/seg3`, etc. This is still less expensive than filtering `n` search calls.

Reply:
I just verified `search` always returns 20 entries starting from an offset.
```
curl  "https://gitlab.com/api/v4/projects/gitlab-org%2Fgitlab/repository/branches?search=%5Efeature"
```

Reply:
What we can do is:

1. implementing proper pagination or
2. using the `get-single-repository-branch` API call you mentioned above and iterate over the subsets of segments in our code, e.g. given the example context URL above:
  a) check if there is a branch called `corneliusludmann/important-issue-1/test-dir-1`, if not ...
  b) check if there is a branch called `corneliusludmann/important-issue-1` (is true so terminate here), in case if not ...
  c) check if there is a branch called `corneliusludmann`.
  or
3. doing the same as (3) but in reverse order (starting with 1 segment, then 2 segments, ...)

The question is, which approach is the best to limit the number of API calls.

(1) would need many API calls if there are a lot of branches starting with the same segment (as `feature/`). We could increase the `perPage` option but still need to implement pagination handling.

(2) would need many API calls if the context URL has a deep file tree. However, I think that most of our context URLs don't have a file tree part at all.

(3) would need some API calls if the branch has some segments. However, the number of segments for branches should be usually small. But something like a `feature/` prefix is quite common. So, it needs pretty often at least 2 calls.

My gut feeling is that with (2), in most of the cases 1 API call is sufficient. But in the worst case (a user has a pretty long path in the context URL) it takes the most API calls.

Probably a combination would be the best idea: First, try if there is a branch with all segments (which is probably in most cases successful) and continue with (3)  in case there is no match (or even better continue with (1) instead ?).

Do you have an opinion/feeling about this, @AlexTugarev?

Reply:
Given that in git you cannot create "prefix" branches, I'd vote for the ascending number of segments to try with. It should be correct in any reasonable way.
Also, if you want to optimize the stuff, add a special handling for the `feature` case and begin with `seg1/seg2` ;-)

Reply:
(3) 👍🏻

Reply:
Regarding pagination: All info I found about the GitLab library we use, it handles pagination behind the scenes already. Newer versions of this library give more control over pagination, though.

However, that also means, if there are a lot of pages that are needed to fetch from GitLab, the library produces a lot of API calls. I'm all in favor of implementing this with pagination as fallback:

> First, try if there is a branch with all segments (which is probably in most cases successful) and continue with (3) in case there is no match (or even better continue with (1) instead ?).

However, that still not explains for sure the root cause of the actual problem.

Reply:
If added:
https://github.com/gitpod-io/gitpod/blob/7785a6c2ca7b58c2ed85d02a13be39b553dad6af/components/server/src/gitlab/gitlab-context-parser.ts#L221

That should be sufficient for most of our cases (and for all prebuilds because they never have a file path in the context URL).

To verify, that the GitLab library we use handles pagination correctly, I added a log entry that tells us the number of fetch branches/tags:

https://github.com/gitpod-io/gitpod/blob/7785a6c2ca7b58c2ed85d02a13be39b553dad6af/components/server/src/gitlab/gitlab-context-parser.ts#L291-L293

If we observe that it's always 20 when it fails, we know for sure that pagination is the problem.

What do you think?

Reply:
👍🏻  for the optimization to start with all segments joined, which might be the dominant case, at least one of them.

just to recap the situation with `feature/9999_new_feature` branches: even if pagination works nicely, it could lead to `(n/20)` API calls, which might still might eat up one sec or two for larger repos/projects. 

I believe a) it will be less API call in average and b) the code will be less verbose if we sort the array of segment prefix' and loop over a single procedure.

Reply:
@AlexTugarev: I'm convinced. Here we are. The PR has been updated.

Unfortunately, the GitLab API answers in some situations with 500 instead of 404 when the branch name is incorrect. We need to check for both. Let's hope, that this will not hide other reasons for that 500 error.

Reply:
Can you post such a case, please? It doesn't sound right. We can file an issue for GitLab.

Reply:
> Can you post such a case, please? It doesn't sound right.

`https://gitlab.com/gp-test-group/gp-test-subgroup/gp-test-project-in-subgroup/-/tree/corneliusludmann` is one

> We can file an issue for GitLab.

Exactly. I have it already on my to-do list.

Reply:
GitLab issue: https://gitlab.com/gitlab-org/gitlab/-/issues/335867

Reply:
🥇  Awesome! Thanks for taking care of this and providing details to GitLab!

Reply:
quick note on the response 500.
I've seen spikes of request times in server's logs. instead of regular ~300ms it took ~5s, which after checking turned out to be the response 500.
the it was a show branch request for `foo`, where `foo` does not exist but `foo/bar` does. this appears to be a real bug in GitLab.